ocsgw/src/main/java/org/ostelco/ocsgw/datasource/protobuf/ProtobufDataSource.kt

```diff
@@ -0,0 +1,169 @@
+package org.ostelco.ocsgw.datasource.protobuf
+
+import org.jdiameter.api.IllegalDiameterStateException
+import org.jdiameter.api.InternalException
+import org.jdiameter.api.OverloadException
+import org.jdiameter.api.RouteException
+import org.jdiameter.api.cca.ServerCCASession
+import org.ostelco.diameter.CreditControlContext
+import org.ostelco.diameter.getLogger
+import org.ostelco.diameter.model.CreditControlAnswer
+import org.ostelco.diameter.model.CreditControlRequest
+import org.ostelco.diameter.model.MultipleServiceCreditControl
+import org.ostelco.diameter.model.ResultCode.DIAMETER_UNABLE_TO_COMPLY
+import org.ostelco.diameter.model.SessionContext
+import org.ostelco.ocs.api.ActivateResponse
+import org.ostelco.ocs.api.CreditControlAnswerInfo
+import org.ostelco.ocs.api.CreditControlRequestInfo
+import org.ostelco.ocs.api.CreditControlRequestType
+import org.ostelco.ocsgw.OcsServer
+import org.ostelco.ocsgw.converter.ProtobufToDiameterConverter
+import org.ostelco.prime.metrics.api.OcsgwAnalyticsReport
+import org.ostelco.prime.metrics.api.User
+import java.util.*
+import java.util.concurrent.ConcurrentHashMap
+
+class ProtobufDataSource {
+
+    private val logger by getLogger()
+
+    private val blocked = Collections.newSetFromMap(ConcurrentHashMap<String, Boolean>())
+
+    private val ccrMap = ConcurrentHashMap<String, CreditControlContext>()
+
+    private val sessionIdMap = ConcurrentHashMap<String, SessionContext>()
+
+    fun handleRequest(context: CreditControlContext): CreditControlRequestInfo? {
+
+        logger.info("[>>] creditControlRequest for {}", context.creditControlRequest.msisdn)
+
+        // FixMe: We should handle conversion errors
+        val creditControlRequestInfo = ProtobufToDiameterConverter.convertRequestToGrpc(context)
+        if (creditControlRequestInfo != null) {
```

Comment:
This could be simplified to :        

if (creditControlRequestInfo != null) {
            ccrMap[context.sessionId] = context
            addToSessionMap(context)
}
return creditControlRequestInfo